pkg/client/record/event.go

```diff
@@ -42,49 +55,82 @@ type EventRecorder interface {
 // your binary. Subsequent calls will be ignored. The return value can be ignored
 // or used to stop recording, if desired.
 func StartRecording(recorder EventRecorder, source api.EventSource) watch.Interface {
+	// Set up our own personal buffer of events so that we can clear out GetEvents'
+	// broadcast channel as quickly as possible to avoid causing the relatively more
+	// important event-producing goroutines from blocking while trying to insert events.
+	eventQueue := make(chan *api.Event, maxQueuedEvents)
+
+	// Run a function in the background that grabs events off the queue and tries
+	// to record them, retrying as appropriate to try to avoid dropping any.
+	go func() {
+		defer util.HandleCrash()
+		for event := range eventQueue {
+			tries := 0
+			for {
+				if recordEvent(recorder, event) {
+					break
+				}
+				tries++
+				if tries >= maxTriesPerEvent {
+					glog.Errorf("Unable to write event '%#v' (retry limit exceeded!)", event)
+					break
+				}
+				sleepDuration := time.Duration(
+					math.Min(maxSleep, minSleep*math.Pow(backoffExp, float64(tries-1))))
+				time.Sleep(wait.Jitter(sleepDuration, 0.5))
+			}
+		}
+	}()
+
+	// Finally, kick off the watcher that takes events from the channel and puts them
+	// onto the queue.
 	return GetEvents(func(event *api.Event) {
 		// Make a copy before modification, because there could be multiple listeners.
 		// Events are safe to copy like this.
 		eventCopy := *event
 		event = &eventCopy
 		event.Source = source
-		try := 0
-		for {
-			try++
-			_, err := recorder.Create(event)
-			if err == nil {
-				break
-			}
-			// If we can't contact the server, then hold everything while we keep trying.
-			// Otherwise, something about the event is malformed and we should abandon it.
-			giveUp := false
-			switch err.(type) {
-			case *client.RequestConstructionError:
-				// We will construct the request the same next time, so don't keep trying.
-				giveUp = true
-			case *errors.StatusError:
-				// This indicates that the server understood and rejected our request.
-				giveUp = true
-			case *errors.UnexpectedObjectError:
-				// We don't expect this; it implies the server's response didn't match a
-				// known pattern. Go ahead and retry.
-			default:
-				// This case includes actual http transport errors. Go ahead and retry.
-			}
-			if giveUp {
-				glog.Errorf("Unable to write event '%#v': '%v' (will not retry!)", event, err)
-				break
-			}
-			if try >= 3 {
-				glog.Errorf("Unable to write event '%#v': '%v' (retry limit exceeded!)", event, err)
-				break
-			}
-			glog.Errorf("Unable to write event: '%v' (will retry in 1 second)", err)
-			time.Sleep(1 * time.Second)
+		// Drop new events rather than old ones because the old ones may contain
+		// some information explaining why everything is so backed up.
+		if len(eventQueue) == maxQueuedEvents {
```

Comment:
Wow, thanks for the code sample. I didn't know you could select based on adding an event to a channel.

I really wasn't too concerned about the race since the only race I see is dropping an event if the channel dropped from full to having a space in between the len() call and the dropping of the event, which should be very uncommon, but this is much nicer.